File: autokeren/checkpoints/manager.py
"""Checkpoint manager — create, load, rewind tool call checkpoints."""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from autokeren.checkpoints.storage import CheckpointStorage


@dataclass
class FileChange:
    path: str
    action: str  # "create", "modify", "delete"
    before: str | None = None
    after: str | None = None


@dataclass
class Checkpoint:
    id: int
    timestamp: float
    tool_name: str
    tool_args: dict[str, Any]
    tool_result: dict[str, Any]
    tool_ok: bool
    file_changes: list[FileChange] = field(default_factory=list)


class CheckpointManager:
    """Manage checkpoints untuk Time-Travel /rewind."""

    def __init__(
        self,
        project_root: Path,
        session_id: str = "default",
        max_checkpoints: int = 50,
        auto_checkpoint: bool = True,
    ) -> None:
        self.project_root = project_root
        self.storage = CheckpointStorage(project_root, session_id, max_checkpoints)
        self.auto_checkpoint = auto_checkpoint
        self._next_id = 1
        self._init_next_id()

    def _init_next_id(self) -> None:
        ids = self.storage.list_ids()
        self._next_id = (max(ids) + 1) if ids else 1

    def _snapshot_file(self, path: str) -> str | None:
        p = self.project_root / path
        if not p.exists() or not p.is_file():
            return None
        try:
            return p.read_text(encoding="utf-8", errors="replace")
        except Exception:
            return None

    def _detect_changes(
        self, tool_name: str, tool_args: dict[str, Any], before_snapshot: dict[str, str | None]
    ) -> list[FileChange]:
        changes: list[FileChange] = []
        if tool_name not in ("write_file", "patch_file"):
            return changes
        path = tool_args.get("path", "")
        if not path:
            return changes
        before = before_snapshot.get(path)
        after = self._snapshot_file(path)
        if before is None and after is not None:
            changes.append(FileChange(path=path, action="create", before=None, after=after))
        elif before is not None and after is not None:
            if before != after:
                changes.append(FileChange(path=path, action="modify", before=before, after=after))
        elif before is not None and after is None:
            changes.append(FileChange(path=path, action="delete", before=before, after=None))
        return changes

    def save(
        self,
        tool_name: str,
        tool_args: dict[str, Any],
        tool_result: dict[str, Any],
        tool_ok: bool,
        before_snapshot: dict[str, str | None] | None = None,
    ) -> Checkpoint:
        if before_snapshot is None:
            before_snapshot = {}
        changes = self._detect_changes(tool_name, tool_args, before_snapshot)
        cp_id = self._next_id
        self._next_id += 1
        now = time.time()
        data = {
            "id": cp_id,
            "timestamp": now,
            "tool_name": tool_name,
            "tool_args": tool_args,
            "tool_result": tool_result,
            "tool_ok": tool_ok,
            "file_changes": [
                {"path": c.path, "action": c.action, "before": c.before, "after": c.after}
                for c in changes
            ],
        }
        self.storage.save(cp_id, data)
        self.storage.save_meta({"next_id": self._next_id})
        ts: float = now
        return Checkpoint(
            id=cp_id,
            timestamp=ts,
            tool_name=tool_name,
            tool_args=tool_args,
            tool_result=tool_result,
            tool_ok=tool_ok,
            file_changes=changes,
        )

    def rewind(self, steps: int = 1) -> list[Checkpoint]:
        if steps < 1:
            steps = 1
        ids = self.storage.list_ids()
        if not ids:
            return []
        to_undo = ids[-steps:] if steps <= len(ids) else ids[:]
        undone: list[Checkpoint] = []
        for cp_id in reversed(to_undo):
            data = self.storage.load(cp_id)
            if data is None:
                continue
            changes = [
                FileChange(
                    path=c["path"],
                    action=c["action"],
                    before=c.get("before"),
                    after=c.get("after"),
                )
                for c in data.get("file_changes", [])
            ]
            cp = Checkpoint(
                id=cp_id,
                timestamp=data["timestamp"],
                tool_name=data["tool_name"],
                tool_args=data.get("tool_args", {}),
                tool_result=data.get("tool_result", {}),
                tool_ok=data.get("tool_ok", True),
                file_changes=changes,
            )
            self._revert_changes(changes)
            self.storage.delete(cp_id)
            undone.append(cp)
        if undone:
            self._next_id = undone[-1].id
            self.storage.save_meta({"next_id": self._next_id})
        return undone

    def _revert_changes(self, changes: list[FileChange]) -> None:
        for change in reversed(changes):
            p = self.project_root / change.path
            if change.action == "create":
                if p.exists():
                    p.unlink()
            elif change.action == "modify":
                if change.before is not None:
                    p.parent.mkdir(parents=True, exist_ok=True)
                    p.write_text(change.before, encoding="utf-8")
            elif change.action == "delete":
                if change.before is not None:
                    p.parent.mkdir(parents=True, exist_ok=True)
                    p.write_text(change.before, encoding="utf-8")

    def list_checkpoints(self) -> list[dict[str, Any]]:
        ids = self.storage.list_ids()
        result = []
        for cp_id in ids:
            data = self.storage.load(cp_id)
            if data is None:
                continue
            result.append({
                "id": cp_id,
                "timestamp": data["timestamp"],
                "tool": data["tool_name"],
                "args": data.get("tool_args", {}),
                "ok": data.get("tool_ok", True),
                "changes": len(data.get("file_changes", [])),
            })
        return result

    def count(self) -> int:
        return self.storage.count()

    def clear(self) -> None:
        self.storage.clear()
        self._next_id = 1
        self.storage.save_meta({"next_id": 1})

    def snapshot_files(self, paths: list[str]) -> dict[str, str | None]:
        return {p: self._snapshot_file(p) for p in paths}

File: internal/checkpoint/manager.go
package checkpoint

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
	"unicode"

	"github.com/autokeren/autokeren/internal/safety"
)

type FileChange struct {
	Path   string  `json:"path"`
	Action string  `json:"action"`
	Before *string `json:"before"`
	After  *string `json:"after"`
}

type Entry struct {
	ID          int            `json:"id"`
	Timestamp   float64        `json:"timestamp"`
	ToolName    string         `json:"tool_name"`
	ToolArgs    map[string]any `json:"tool_args"`
	ToolResult  map[string]any `json:"tool_result"`
	ToolOK      bool           `json:"tool_ok"`
	FileChanges []FileChange   `json:"file_changes"`
}

type Manager struct {
	root           string
	directory      string
	maxCheckpoints int
	auto           bool
	nextID         int
}

func New(root, sessionID string, maxCheckpoints int, auto bool) (*Manager, error) {
	if strings.TrimSpace(sessionID) == "" {
		sessionID = "default"
	}
	if maxCheckpoints <= 0 {
		maxCheckpoints = 50
	}
	resolved := resolveRoot(root)
	directory, err := checkpointDirectory(resolved, sessionID)
	if err != nil {
		return nil, err
	}
	manager := &Manager{root: resolved, directory: directory, maxCheckpoints: maxCheckpoints, auto: auto, nextID: 1}
	if ids, err := manager.ids(); err == nil && len(ids) > 0 {
		manager.nextID = ids[len(ids)-1] + 1
	}
	return manager, nil
}

func (m *Manager) Enabled() bool {
	return m != nil && m.auto
}

func (m *Manager) Directory() string {
	if m == nil {
		return ""
	}
	return m.directory
}

func (m *Manager) Snapshot(toolName string, args map[string]any) map[string]*string {
	if m == nil || (toolName != "write_file" && toolName != "patch_file") {
		return nil
	}
	path, _ := args["path"].(string)
	if path == "" {
		return nil
	}
	content, exists := m.readProjectFile(path)
	if !exists {
		return map[string]*string{path: nil}
	}
	return map[string]*string{path: &content}
}

func (m *Manager) Save(toolName string, args map[string]any, result map[string]any, ok bool, before map[string]*string) (Entry, error) {
	if m == nil {
		return Entry{}, nil
	}
	changes := m.detectChanges(toolName, args, before)
	entry := Entry{ID: m.nextID, Timestamp: float64(time.Now().UnixNano()) / 1e9, ToolName: toolName, ToolArgs: args, ToolResult: result, ToolOK: ok, FileChanges: changes}
	data, err := json.MarshalIndent(entry, "", "  ")
	if err != nil {
		return Entry{}, fmt.Errorf("serialisasi checkpoint: %w", err)
	}
	if err := writeFileAtomic(m.entryPath(entry.ID), data); err != nil {
		return Entry{}, fmt.Errorf("simpan checkpoint: %w", err)
	}
	m.nextID++
	if err := m.rotate(); err != nil {
		return Entry{}, err
	}
	meta, _ := json.MarshalIndent(map[string]int{"next_id": m.nextID}, "", "  ")
	if err := writeFileAtomic(filepath.Join(m.directory, "meta.json"), meta); err != nil {
		return Entry{}, fmt.Errorf("simpan metadata checkpoint: %w", err)
	}
	return entry, nil
}

func (m *Manager) Rewind(steps int) ([]Entry, error) {
	if m == nil {
		return nil, errors.New("time-travel tidak diaktifkan")
	}
	if steps < 1 {
		steps = 1
	}
	ids, err := m.ids()
	if err != nil {
		return nil, err
	}
	if len(ids) == 0 {
		return nil, nil
	}
	if steps > len(ids) {
		steps = len(ids)
	}
	targets := ids[len(ids)-steps:]
	undone := make([]Entry, 0, len(targets))
	for index := len(targets) - 1; index >= 0; index-- {
		entry, err := m.load(targets[index])
		if err != nil {
			return undone, err
		}
		if err := m.revert(entry.FileChanges); err != nil {
			return undone, err
		}
		if err := os.Remove(m.entryPath(entry.ID)); err != nil && !os.IsNotExist(err) {
			return undone, err
		}
		undone = append(undone, entry)
	}
	if len(undone) > 0 {
		m.nextID = undone[len(undone)-1].ID
		meta, _ := json.MarshalIndent(map[string]int{"next_id": m.nextID}, "", "  ")
		if err := writeFileAtomic(filepath.Join(m.directory, "meta.json"), meta); err != nil {
			return undone, err
		}
	}
	return undone, nil
}

func (m *Manager) List() ([]Entry, error) {
	if m == nil {
		return nil, nil
	}
	ids, err := m.ids()
	if err != nil {
		return nil, err
	}
	entries := make([]Entry, 0, len(ids))
	for _, id := range ids {
		entry, err := m.load(id)
		if err != nil {
			return nil, err
		}
		entries = append(entries, entry)
	}
	return entries, nil
}

func (m *Manager) Count() int {
	entries, err := m.List()
	if err != nil {
		return 0
	}
	return len(entries)
}

func (m *Manager) detectChanges(toolName string, args map[string]any, before map[string]*string) []FileChange {
	if toolName != "write_file" && toolName != "patch_file" {
		return nil
	}
	path, _ := args["path"].(string)
	if path == "" {
		return nil
	}
	prior, tracked := before[path]
	if !tracked {
		return nil
	}
	afterText, exists := m.readProjectFile(path)
	if prior == nil && exists {
		return []FileChange{{Path: path, Action: "create", After: &afterText}}
	}
	if prior != nil && exists && *prior != afterText {
		return []FileChange{{Path: path, Action: "modify", Before: prior, After: &afterText}}
	}
	if prior != nil && !exists {
		return []FileChange{{Path: path, Action: "delete", Before: prior}}
	}
	return nil
}

func (m *Manager) revert(changes []FileChange) error {
	for index := len(changes) - 1; index >= 0; index-- {
		change := changes[index]
		path, err := projectPath(m.root, change.Path)
		if err != nil {
			return err
		}
		switch change.Action {
		case "create":
			if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
				return err
			}
		case "modify", "delete":
			if change.Before != nil {
				if err := writeFileAtomic(path, []byte(*change.Before)); err != nil {
					return err
				}
			}
		}
	}
	return nil
}

func (m *Manager) readProjectFile(path string) (string, bool) {
	target, err := projectPath(m.root, path)
	if err != nil {
		return "", false
	}
	data, err := os.ReadFile(target)
	if err != nil {
		return "", false
	}
	return string(data), true
}

func (m *Manager) load(id int) (Entry, error) {
	data, err := os.ReadFile(m.entryPath(id))
	if err != nil {
		return Entry{}, err
	}
	var entry Entry
	if err := json.Unmarshal(data, &entry); err != nil {
		return Entry{}, fmt.Errorf("baca checkpoint %d: %w", id, err)
	}
	return entry, nil
}

func (m *Manager) ids() ([]int, error) {
	entries, err := os.ReadDir(m.directory)
	if err != nil {
		return nil, err
	}
	ids := make([]int, 0)
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".json" || entry.Name() == "meta.json" {
			continue
		}
		var id int
		if _, err := fmt.Sscanf(strings.TrimSuffix(entry.Name(), ".json"), "%d", &id); err == nil && id > 0 {
			ids = append(ids, id)
		}
	}
	sort.Ints(ids)
	return ids, nil
}

func (m *Manager) rotate() error {
	ids, err := m.ids()
	if err != nil {
		return err
	}
	for len(ids) > m.maxCheckpoints {
		if err := os.Remove(m.entryPath(ids[0])); err != nil {
			return err
		}
		ids = ids[1:]
	}
	return nil
}

func (m *Manager) entryPath(id int) string {
	return filepath.Join(m.directory, fmt.Sprintf("%04d.json", id))
}

func checkpointDirectory(root, sessionID string) (string, error) {
	primary := filepath.Join(root, ".ak-checkpoints", "session-"+sessionID)
	if err := testDirectory(primary); err == nil {
		return primary, nil
	}
	fallback := filepath.Join(configBase(), "projects", projectSlug(root), ".ak-checkpoints", "session-"+sessionID)
	if err := testDirectory(fallback); err != nil {
		return "", err
	}
	return fallback, nil
}

func testDirectory(directory string) error {
	if err := os.MkdirAll(directory, 0o700); err != nil {
		return err
	}
	file, err := os.CreateTemp(directory, ".write-test-")
	if err != nil {
		return err
	}
	name := file.Name()
	if closeErr := file.Close(); closeErr != nil {
		return closeErr
	}
	return os.Remove(name)
}

func writeFileAtomic(path string, data []byte) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	file, err := os.CreateTemp(filepath.Dir(path), ".checkpoint-")
	if err != nil {
		return err
	}
	temporary := file.Name()
	defer os.Remove(temporary)
	if err := file.Chmod(0o600); err != nil {
		file.Close()
		return err
	}
	if _, err := file.Write(data); err != nil {
		file.Close()
		return err
	}
	if err := file.Close(); err != nil {
		return err
	}
	return os.Rename(temporary, path)
}

func projectPath(root, requested string) (string, error) {
	return safety.ProjectPath(root, requested)
}

func resolveRoot(root string) string {
	resolved, err := filepath.Abs(root)
	if err != nil {
		resolved = root
	}
	if realPath, err := filepath.EvalSymlinks(resolved); err == nil {
		return realPath
	}
	return filepath.Clean(resolved)
}

func configBase() string {
	if base := strings.TrimSpace(os.Getenv("AUTOKEREN_CONFIG_DIR")); base != "" {
		return base
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return filepath.Join(".", ".config", "autokeren")
	}
	return filepath.Join(home, ".config", "autokeren")
}

func projectSlug(root string) string {
	name := strings.Trim(strings.Map(func(r rune) rune {
		if unicode.IsLetter(r) || unicode.IsNumber(r) || r == '_' || r == '-' || r == '.' {
			return r
		}
		return '_'
	}, filepath.Base(root)), "._")
	if name == "" {
		name = "default"
	}
	sum := md5.Sum([]byte(root))
	return name + "-" + hex.EncodeToString(sum[:])[:8]
}
